graphql_compiler/schema/__init__.py

```diff
@@ -340,36 +340,47 @@ def _serialize_date(value: Any) -> str:
 
 def _parse_date_value(value: Any) -> date:
     """Deserialize a Date object from its proper ISO-8601 representation."""
-    return arrow.get(value, "YYYY-MM-DD").date()
+    if type(value) == date:
+        # We prefer exact type equality instead of isinstance() because datetime objects
+        # are subclasses of date but are not interchangeable for dates for our purposes.
+        return value
+    elif isinstance(value, str):
+        return arrow.get(value, "YYYY-MM-DD").date()
+    else:
+        raise ValueError(
+            f"Expected a date object or its ISO-8601 string representation. "
+            f"Got {value} of type {type(value)} instead."
+        )
 
 
 def _serialize_datetime(value: Any) -> str:
     """Serialize a DateTime object to its proper ISO-8601 representation."""
-    # Python datetime.datetime is a subclass of datetime.date, but in this case, the two are not
-    # interchangeable. Rather than using isinstance, we will therefore check for exact type
-    # equality.
-    #
     # We don't allow Arrow objects as input since it seems that Arrow objects are always tz aware.
     # This is supported by the fact that the `.naive` Arrow method returns a datetime object instead
     # of an Arrow object.
-    if type(value) == datetime and value.tzinfo is None:
+    if isinstance(value, datetime) and value.tzinfo is None:
         return value.isoformat()
     else:
         raise ValueError(
-            f"Expected a timezone naive datetime object. Got {value} of type {type(value)} instead."
+            f"Expected a timezone-naive datetime object. Got {value} of type {type(value)} instead."
         )
 
 
 def _parse_datetime_value(value: Any) -> datetime:
     """Deserialize a DateTime object from its proper ISO-8601 representation."""
-    # attempt to parse with microsecond information
-    try:
-        arrow_result = arrow.get(value, "YYYY-MM-DDTHH:mm:ss")
-    except arrow.parser.ParserMatchError:
-        arrow_result = arrow.get(value, "YYYY-MM-DDTHH:mm:ss.S")
-
-    # arrow parses datetime naive strings into Arrow objects with a UTC timezone.
-    return arrow_result.naive
+    if isinstance(value, datetime) and value.tzinfo is None:
```

Comment:
Should this also be checking for exact type equality like with `date` or do we expect subclasses of `datetime` to appear at some point? If the latter, does that mean we're trusting whatever library we choose (`arrow`, `ciso8601`, etc) to parse the subclasses correctly?

Reply:
I removed our dependency on `arrow` and switched everything over to `ciso8601` which returns "real" Python `datetime` objects, so while I don't expect subclasses of `datetime` to appear, I don't think we necessarily need to prevent them either.

The reason we check for exact type equality in the `date` case is to prevent implicit loss of precision: since `datetime` is technically a subclass of `date`, `isinstance(some_datetime, date)` would return true but when it's used as a date, the time and time zone components would be implicitly lost. In the opposite direction, there is no associated loss of precision (the conversion is widening, not narrowing), so this is not really a concern. Additionally, `isinstance(some_date, datetime)` returns `False`, which is nice and as expected 😄